lib/tbot/config/destination_kubernetes_secret.go

```diff
@@ -0,0 +1,222 @@
+	// need to worry about writing/reading from a nil map.
+	if secret.Data == nil {
+		secret.Data = map[string][]byte{}
+	}
+	return secret, nil
+}
+
+func (dks *DestinationKubernetesSecret) secretTemplate() *corev1.Secret {
+	return &corev1.Secret{
+		Type: corev1.SecretTypeOpaque,
+		ObjectMeta: v1.ObjectMeta{
+			Name:      dks.Name,
+			Namespace: dks.namespace,
+		},
+		Data: map[string][]byte{},
+	}
+}
+
+func (dks *DestinationKubernetesSecret) upsertSecret(ctx context.Context, secret *corev1.Secret) error {
+	apply := applyconfigv1.Secret(dks.Name, dks.namespace).
+		WithData(secret.Data).
+		WithResourceVersion(secret.ResourceVersion).
+		WithType(secret.Type)
+
+	_, err := dks.k8s.CoreV1().Secrets(dks.namespace).Apply(ctx, apply, v1.ApplyOptions{
+		FieldManager: dks.fieldManager(),
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (dks *DestinationKubernetesSecret) Verify(_ []string) error {
+	return nil
+}
+
+func (dks *DestinationKubernetesSecret) TryLock() (func() error, error) {
+	// No locking support currently implemented. Users will need to be cautious
+	// to not point two tbots to the same secret.
+	return func() error { return nil }, nil
+}
+
+func (dks *DestinationKubernetesSecret) CheckAndSetDefaults() error {
+	if dks.Name == "" {
+		return trace.BadParameter("name must not be empty")
+	}
+
+	return nil
+}
+
+func (dks *DestinationKubernetesSecret) Init(ctx context.Context, subdirs []string) error {
+	dks.mu.Lock()
+	defer dks.mu.Unlock()
+	if dks.initialized == true {
+		return trace.BadParameter("destination has already been initialized")
+	}
+
+	if dks.namespace == "" {
+		dks.namespace = os.Getenv(kubernetesNamespaceEnv)
+		if dks.namespace == "" {
+			return trace.BadParameter("unable to detect namespace from %s environment variable", kubernetesNamespaceEnv)
+		}
+	}
+
+	if len(subdirs) > 0 {
+		return trace.BadParameter("kubernetes_secret destination does not support subdirectories")
+	}
+
+	// If no k8s client is injected, we attempt to create one from the
+	// environment.
+	if dks.k8s == nil {
+		// BuildConfigFromFlags falls back to InClusterConfig if both params
+		// are empty. This means KUBECONFIG takes precedence.
+		clientCfg, err := clientcmd.BuildConfigFromFlags("", os.Getenv("KUBECONFIG"))
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		dks.k8s, err = kubernetes.NewForConfig(clientCfg)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	// Try and read the secret and then write to it. This lets us fail early
+	// and clearly if the appropriate permissions are missing
+	secret, err := dks.getSecret(ctx)
+	if err != nil {
+		if !kubeerrors.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		log.WithFields(logrus.Fields{
+			"secret_name":      dks.Name,
+			"secret_namespace": dks.namespace,
+		}).Debug("Did not find an existing Kubernetes secret. One will be created.")
+		secret = dks.secretTemplate()
+	}
+	if err := dks.upsertSecret(ctx, secret); err != nil {
```

Comment:
Can we run this step with `dry-run: server`?

The issue is if the pod that consumes the secret is waiting for it, once `tbot` creates an empty secret during `Init`, the pod will start and probably fail because:
- the secret is empty 
- secret update takes some time to propagate and the user app will fail.

Running the init phase with dry-run blocks the pod from being scheduled until the secret is actually created.